### PR TITLE
[Testcase Refactoring] Refactor test_c10d_fault_tolerance with instantiate_device_type_tests and hw-classification

### PR DESCRIPTION
**Pull request by PlatformLC**
## Summary
Refactor `test_c10d_fault_tolerance` to use `instantiate_device_type_tests` + `hw-classification`, and gate reconfigure support through the #190846 capability framework (stacks on #190846). Each test method is decorated with `@requires_capabilities`, so a variant whose backend lacks reconfigure skips at run time, while the original `assertTrue` contract is preserved as a second layer.

## Changes

**`torch/testing/_internal/common_device_type.py`** (extends #190846's framework):
- Add `Capability.collective.{reconfigure, work_result}` to the `Capability` namespace.
- `CPUTestBase._capabilities`: declare `reconfigure = dist.is_available()` (gloo, the default CPU backend, supports reconfigure and is built whenever distributed is) and `work_result = False` (gloo does not report `WorkResult`).
- `CUDATestBase._capabilities`: declare both `reconfigure` and `work_result` via a nested `nccl2_available()` helper (`dist.is_available() and dist.is_backend_available("nccl2")`) — nccl2 is the reconfigure-capable CUDA backend (the default `"nccl"` is not) and the only backend that reports `WorkResult`.

**`test/distributed/test_c10d_fault_tolerance.py`**:
- Replace the `_make_fault_tolerance_test_class` factory + `globals()` registration loop with `instantiate_device_type_tests(FaultToleranceTest, globals(), except_for=["hpu", "xpu"])`; classify the reconfigure collective tests as `HardwareClassification.ACCELERATOR`.
- Decorate each test method with `@requires_capabilities`: the 13 reconfigure tests declare `Capability.collective.reconfigure`; the 2 work-result tests declare both `reconfigure` and `work_result`. A variant whose device base reports the capability as missing/unsupported raises `SkipTest` at run time, before `_init_reconfigurable_pg`.
- The `assertTrue(dist._supports_reconfigure())` / `assertTrue(backend.supports_reconfigure)` contract in `_init_reconfigurable_pg` is **preserved** as a second layer: it catches a backend that lacks reconfigure even if a variant reaches init (e.g. if an OOT base later declares the capability before the backend actually supports it).
- Override `MultiProcessTestCase.run_test` to re-invoke `type(self).setUpClass()` before the parent's `run_test`. Workers run the test method directly via `run_test`, skipping `setUpClass`, so `PrivateUse1TestBase` would never translate the `"privateuse1"` device_type token to the real backend name in the worker; the override makes class-level device setup apply in the worker like in the main process. `device`/`backend_name`/`setUp` then read `self.device_type` directly (no `_resolved_device_type` shim).
- Generalize the device setup: `set_device` → `torch.accelerator.set_device_index(self.rank)` (was cuda-only); the 3-device gate uses `torch.accelerator.device_count()`.
- The 15 test method bodies and per-test PG lifecycle are unchanged; the two pure-logic contract classes (`ReconfigureContractTest`, `BackendCapabilityContractTest`) are labeled `GENERIC` and not instantiated.

## Motivation
Part of the test case refactoring initiative (#185590) to label tests by hardware classification and generate per-device variants via `instantiate_device_type_tests`, so reconfigure (fault-tolerance) collective behavior is exercised across reconfigure-capable accelerator backends rather than hand-rolled per-backend.

## Notes
- **Depends on #190846** (the capability-based test-skip framework). This PR extends that framework with the `collective.reconfigure` / `collective.work_result` capabilities and is the first consumer.

### CODE REVIEW

File: test/cpp_extensions/open_registration_extension/torch_openreg/tests/test_testing.py [hunk 1]
```diff
@@ -7,13 +7,15 @@
 import torch
```
Comment (PlatformLC) [inline test/cpp_extensions/open_registration_extension/torch_openreg/tests/test_testing.py:8]: this file's change if from https://github.com/pytorch/pytorch/pull/190846 capability framework. Not part of this PR.

File: torch/testing/_internal/common_device_type.py [hunk 2]
```diff
@@ -755,6 +839,40 @@ class CUDATestBase(DeviceTypeTestBase):
...
+            Capability.lib: {
+                Capability.lib.triton: lambda: has_triton(),
+            },
+            Capability.attention: {
+                Capability.attention.flash_attention: lambda: PLATFORM_SUPPORTS_FLASH_ATTENTION,
+                Capability.attention.mem_efficient_attention: lambda: PLATFORM_SUPPORTS_MEM_EFF_ATTENTION,
+            },
+            # nccl2 is the reconfigure-capable CUDA backend (default "nccl" is
+            # not) and is the only backend that reports WorkResult.
+            Capability.collective: {
```
Comment (PlatformLC) [inline torch/testing/_internal/common_device_type.py]: Only `Capability.collective` related change is part of this PR, others are from #190846 capability framework.

File: test/distributed/test_c10d_fault_tolerance.py [hunk 3]
```diff
@@ -370,36 +411,15 @@ def test_reconfigure_timeout_is_retryable(self):
...
-            not TEST_CUDA or torch.cuda.device_count() < 3,
-            "fault tolerance CUDA tests require at least 3 GPUs",
-        )(cls)
-    return cls
 
-
-for backend in FAULT_TOLERANCE_BACKENDS:
-    class_name = f"{backend.name.capitalize()}FaultToleranceTest"
-    globals()[class_name] = _make_fault_tolerance_test_class(backend)
+instantiate_device_type_tests(FaultToleranceTest, globals(), except_for=["hpu", "xpu"])
```
Comment (Niran814804102) [inline test/distributed/test_c10d_fault_tolerance.py:413]: noticed that this case was specifically for CPU and CUDA. 
Regarding the "except for" result, will it introduce new device types, such as MPS?  Do we have any result of the case on MPS been verified?

**Review by Niran814804102 — APPROVED** [review]
Thanks for submitting. Looks like the test framework changes aren't included yet. Before merging, please first merge the dependent pr.

File: test/distributed/test_c10d_fault_tolerance.py (hunk 3 above)
Comment (PlatformLC) [reply test/distributed/test_c10d_fault_tolerance.py:413]: MPS will not be included only if  `allow_mps` is set to `True` intentionally.

**Review by Niran814804102 — APPROVED** [review]
LGTM
